src/VisualStudio/Core/Def/Copilot/VisualStudioCopilotCodeAnalysisService.cs

```diff
@@ -0,0 +1,215 @@
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Settings;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Copilot;
+
+[ExportWorkspaceService(typeof(ICopilotCodeAnalysisService)), Shared]
+internal sealed class VisualStudioCopilotCodeAnalysisService : ICopilotCodeAnalysisService
+{
+    private readonly CopilotDiagnosticUpdateSource _copilotDiagnosticUpdateSource;
+    private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+    private readonly IGlobalOptionService _globalOptions;
+    private readonly ConditionalWeakTable<Document, ConcurrentDictionary<string, (ImmutableArray<Diagnostic> Diagnostics, bool IsCompleteResult)>> _diagnosticsCache = new();
+    private ISettingsManager? _settingsManager;
+    private bool _executingAnalysis;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public VisualStudioCopilotCodeAnalysisService(
+        CopilotDiagnosticUpdateSource copilotDiagnosticUpdateSource,
+        IDiagnosticsRefresher diagnosticsRefresher,
+        IGlobalOptionService globalOptions,
+        IThreadingContext threadingContext,
+        IVsService<SVsSettingsPersistenceManager, ISettingsManager> settingsManagerService)
+    {
+        _copilotDiagnosticUpdateSource = copilotDiagnosticUpdateSource;
+        _diagnosticsRefresher = diagnosticsRefresher;
+        _globalOptions = globalOptions;
+        InitializeAsync(settingsManagerService, threadingContext.DisposalToken).Forget();
+    }
+
+    private async Task InitializeAsync(IVsService<SVsSettingsPersistenceManager, ISettingsManager> settingsManagerService, CancellationToken cancellationToken)
+    {
+        _settingsManager = await settingsManagerService.GetValueAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    public bool IsRefineOptionEnabled(Document document)
+        => IsCopilotOptionEnabled(document, "EnableCSharpRefineQuickActionSuggestion");
+
+    public bool IsCodeAnalysisOptionEnabled(Document document)
+        => IsCopilotOptionEnabled(document, "EnableCSharpCodeAnalysis");
+
+    private bool IsCopilotOptionEnabled(Document document, string optionName)
+    {
+        // The bool setting is persisted as 0=None, 1=True, 2=False, so it needs to be retrieved as an int.
+        if (_settingsManager?.TryGetValue($"Microsoft.VisualStudio.Conversations.{optionName}", out int isEnabled) != GetValueResult.Success)
+            return false;
+
+        return isEnabled == 1 && document.GetLanguageService<ICopilotAnalyzer>() is not null;
+    }
+
+    private static bool IsLspPullDiagnostics(IGlobalOptionService globalOptions)
+        => globalOptions.IsLspPullDiagnostics(InternalDiagnosticsOptionsStorage.NormalDiagnosticMode);
```

Comment:
note: this is all going away.  so this needs to work just with lsp pull and nothing else.